'minio version -h' crashes

```
$ minio version -h
Unknown flag. ‘-h’

panic: template: help:6: unexpected EOF

goroutine 1 [running]:
panic(0xbe6360, 0xc8203de430)
        /home/vadmeste/work/go/src/runtime/panic.go:481 +0x3e6
text/template.Must(0x0, 0x7fbe9a44e000, 0xc8203de430, 0x0)
        /home/vadmeste/work/go/src/text/template/helper.go:23 +0x4b
github.com/minio/minio/vendor/github.com/minio/cli.(*App).Run.func2(0xe7a540, 0x55, 0xcec3c0, 0xc82011c0c0)
        /home/vadmeste/work/gospace/src/github.com/minio/minio/vendor/github.com/minio/cli/app.go:136 +0x451
github.com/minio/minio/vendor/github.com/minio/cli.ShowCommandHelp(0xc8201c0000, 0xd42820, 0x7)
        /home/vadmeste/work/gospace/src/github.com/minio/minio/vendor/github.com/minio/cli/help.go:182 +0x852
github.com/minio/minio/vendor/github.com/minio/cli.Command.Run(0xd42820, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, 0xd4ca80, 0xe, 0x0, ...)
        /home/vadmeste/work/gospace/src/github.com/minio/minio/vendor/github.com/minio/cli/command.go:94 +0xd82
github.com/minio/minio/vendor/github.com/minio/cli.(*App).Run(0xc820109320, 0xc8200760c0, 0x3, 0x3, 0x0, 0x0)
        /home/vadmeste/work/gospace/src/github.com/minio/minio/vendor/github.com/minio/cli/app.go:186 +0x476
github.com/minio/minio/vendor/github.com/minio/cli.(*App).RunAndExitOnError(0xc820109320)
        /home/vadmeste/work/gospace/src/github.com/minio/minio/vendor/github.com/minio/cli/app.go:197 +0x45
github.com/minio/minio/cmd.Main()
        /home/vadmeste/work/gospace/src/github.com/minio/minio/cmd/main.go:198 +0x1f8
main.main()
        /home/vadmeste/work/gospace/src/github.com/minio/minio/main.go:6 +0x14
```

--- a/cmd/commands.go
+++ b/cmd/commands.go
@@ -24,19 +24,6 @@ var commands = []cli.Command{}
 // Collection of minio commands currently supported in a trie tree.
 var commandsTree = newTrie()
 
-// Collection of minio flags currently supported.
-var globalFlags = []cli.Flag{
-	cli.StringFlag{
-		Name:  "config-dir, C",
-		Value: mustGetConfigPath(),
-		Usage: "Path to configuration folder.",
-	},
-	cli.BoolFlag{
-		Name:  "quiet",
-		Usage: "Suppress chatty output.",
-	},
-}
-
 // registerCommand registers a cli command.
 func registerCommand(command cli.Command) {
 	commands = append(commands, command)

--- a/cmd/control-heal-main.go
+++ b/cmd/control-heal-main.go
@@ -29,11 +29,16 @@ var healCmd = cli.Command{
 	Name:   "heal",
 	Usage:  "To heal objects.",
 	Action: healControl,
+	Flags:  globalFlags,
 	CustomHelpTemplate: `NAME:
   minio control {{.Name}} - {{.Usage}}
 
 USAGE:
   minio control {{.Name}}
+
+FLAGS:
+  {{range .Flags}}{{.}}
+  {{end}}
 
 EXAMPLES:
   1. Heal an object.

--- a/cmd/control-lock-main.go
+++ b/cmd/control-lock-main.go
@@ -97,11 +97,16 @@ var lockCmd = cli.Command{
 	Name:   "lock",
 	Usage:  "info about the locks in the node.",
 	Action: lockControl,
+	Flags:  globalFlags,
 	CustomHelpTemplate: `NAME:
   minio control {{.Name}} - {{.Usage}}
 
 USAGE:
   minio control {{.Name}} http://localhost:9000/
+
+FLAGS:
+  {{range .Flags}}{{.}}
+  {{end}}
 
 EAMPLES:
   1. Get all the info about the blocked/held locks in the node:

--- a/cmd/control-main.go
+++ b/cmd/control-main.go
@@ -22,6 +22,7 @@ import "github.com/minio/cli"
 var controlCmd = cli.Command{
 	Name:   "control",
 	Usage:  "Control and manage minio server.",
+	Flags:  globalFlags,
 	Action: mainControl,
 	Subcommands: []cli.Command{
 		lockCmd,

--- a/cmd/control-shutdown-main.go
+++ b/cmd/control-shutdown-main.go
@@ -23,21 +23,27 @@ import (
 	"github.com/minio/cli"
 )
 
+var shutdownFlags = []cli.Flag{
+	cli.BoolFlag{
+		Name:  "restart",
+		Usage: "Restart the server.",
+	},
+}
+
 var shutdownCmd = cli.Command{
 	Name:   "shutdown",
 	Usage:  "Shutdown or restart the server.",
 	Action: shutdownControl,
-	Flags: []cli.Flag{
-		cli.BoolFlag{
-			Name:  "restart",
-			Usage: "Restart the server.",
-		},
-	},
+	Flags:  append(shutdownFlags, globalFlags...),
 	CustomHelpTemplate: `NAME:
   minio control {{.Name}} - {{.Usage}}
 
 USAGE:
   minio control {{.Name}} http://localhost:9000/
+
+FLAGS:
+  {{range .Flags}}{{.}}
+  {{end}}
 
 EXAMPLES:
   1. Shutdown the server:

--- a/cmd/main.go
+++ b/cmd/main.go
@@ -28,10 +28,19 @@ import (
 
 var (
 	// global flags for minio.
-	minioFlags = []cli.Flag{
+	globalFlags = []cli.Flag{
 		cli.BoolFlag{
 			Name:  "help, h",
 			Usage: "Show help.",
+		},
+		cli.StringFlag{
+			Name:  "config-dir, C",
+			Value: mustGetConfigPath(),
+			Usage: "Path to configuration folder.",
+		},
+		cli.BoolFlag{
+			Name:  "quiet",
+			Usage: "Suppress chatty output.",
 		},
 	}
 )
@@ -116,7 +125,7 @@ func registerApp() *cli.App {
 	app.Author = "Minio.io"
 	app.Usage = "Cloud Storage Server."
 	app.Description = `Minio is an Amazon S3 compatible object storage server. Use it to store photos, videos, VMs, containers, log files, or any blob of data as objects.`
-	app.Flags = append(minioFlags, globalFlags...)
+	app.Flags = globalFlags
 	app.Commands = commands
 	app.CustomAppHelpTemplate = minioHelpTemplate
 	app.CommandNotFound = func(ctx *cli.Context, command string) {

--- a/cmd/server-main.go
+++ b/cmd/server-main.go
@@ -29,20 +29,23 @@ import (
 )
 
 var srvConfig serverCmdConfig
-var serverCmd = cli.Command{
-	Name:  "server",
-	Usage: "Start object storage server.",
-	Flags: []cli.Flag{
-		cli.StringFlag{
-			Name:  "address",
-			Value: ":9000",
-			Usage: "Specify custom server \"ADDRESS:PORT\", defaults to \":9000\".",
-		},
-		cli.StringFlag{
-			Name:  "ignore-disks",
-			Usage: "Specify comma separated list of disks that are offline.",
-		},
+
+var serverFlags = []cli.Flag{
+	cli.StringFlag{
+		Name:  "address",
+		Value: ":9000",
+		Usage: "Specify custom server \"ADDRESS:PORT\", defaults to \":9000\".",
 	},
+	cli.StringFlag{
+		Name:  "ignore-disks",
+		Usage: "Specify comma separated list of disks that are offline.",
+	},
+}
+
+var serverCmd = cli.Command{
+	Name:   "server",
+	Usage:  "Start object storage server.",
+	Flags:  append(serverFlags, globalFlags...),
 	Action: serverMain,
 	CustomHelpTemplate: `NAME:
   minio {{.Name}} - {{.Usage}}

--- a/cmd/update-main.go
+++ b/cmd/update-main.go
@@ -34,10 +34,6 @@ import (
 var (
 	updateFlags = []cli.Flag{
 		cli.BoolFlag{
-			Name:  "help, h",
-			Usage: "Help for update.",
-		},
-		cli.BoolFlag{
 			Name:  "experimental, E",
 			Usage: "Check experimental update.",
 		},
@@ -49,7 +45,7 @@ var updateCmd = cli.Command{
 	Name:   "update",
 	Usage:  "Check for a new software update.",
 	Action: mainUpdate,
-	Flags:  updateFlags,
+	Flags:  append(updateFlags, globalFlags...),
 	CustomHelpTemplate: `Name:
    minio {{.Name}} - {{.Usage}}
 

--- a/cmd/version-main.go
+++ b/cmd/version-main.go
@@ -25,15 +25,25 @@ var versionCmd = cli.Command{
 	Name:   "version",
 	Usage:  "Print version.",
 	Action: mainVersion,
+	Flags:  globalFlags,
 	CustomHelpTemplate: `NAME:
    minio {{.Name}} - {{.Usage}}
 
 USAGE:
-   minio {{.Name}} {{if .Description}}
+   minio {{.Name}}
+
+FLAGS:
+  {{range .Flags}}{{.}}
+  {{end}}
+
 `,
 }
 
 func mainVersion(ctx *cli.Context) {
+	if len(ctx.Args()) != 0 {
+		cli.ShowCommandHelpAndExit(ctx, "version", 1)
+	}
+
 	console.Println("Version: " + Version)
 	console.Println("Release-Tag: " + ReleaseTag)
 	console.Println("Commit-ID: " + CommitID)

